{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2a905c9-679d-4a0d-b067-1ee015267632",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import argparse\n",
    "\n",
    "# parser = argparse.ArgumentParser(description='Supply reference fasta and bam file')\n",
    "# parser.add_argument('ref',\n",
    "#                     help='reference fasta')\n",
    "# parser.add_argument('bam',\n",
    "#                     help='bam file')\n",
    "\n",
    "\n",
    "# args = parser.parse_args()\n",
    "# ref_fname = args.ref\n",
    "# bam_fname = args.bam\n",
    "ref_fname = 'data/reference/refchrm.fa'\n",
    "bam_fname = 'test/in10.bam'\n",
    "genomes_fname ='data/genomes/311humans.fasta'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e5fec4e-32d8-4cc2-9086-2618191e076b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from collections import Counter\n",
    "import pysam\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "from IPython.display import clear_output\n",
    "from scipy.special import binom\n",
    "import scipy.stats as st\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import subprocess\n",
    "from multiprocess import Pool\n",
    "%load_ext cython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51868805-f10f-4930-acd2-55c8f479ab6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Cython part"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c172378a-ec05-4ce7-b954-4194af606f81",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%cython -a --compile-args=-O3\n",
    "from tqdm import tqdm\n",
    "import pysam\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from collections import Counter\n",
    "cimport cython\n",
    "import numpy as np\n",
    "from cython.parallel import prange\n",
    "from libc.math cimport pow\n",
    "from scipy.special import binom\n",
    "# from mc_lib.rndm cimport RndmWrapper\n",
    "\n",
    "\n",
    "def get_num_reads(str bam_fname):\n",
    "    samfile = pysam.AlignmentFile(bam_fname, \"rb\" )\n",
    "    num_reads = 0\n",
    "    for read in samfile.fetch('chrM'):\n",
    "        if not read.is_mapped:\n",
    "            continue\n",
    "        num_reads += 1\n",
    "    samfile.close()\n",
    "    return num_reads\n",
    "\n",
    "def bam2consensus(\n",
    "        ref_fname, bam_fname, double ac_threshold=0, double af_threshold=0):\n",
    "\n",
    "    cdef str consensus = ''\n",
    "    cdef int max_count, total_cgeount\n",
    "    cdef str allele\n",
    "    cdef str max_allele\n",
    "    for record in SeqIO.parse(ref_fname, \"fasta\"):\n",
    "        assert record.id == 'chrM'\n",
    "        # consensus = \"N\" * len(record)\n",
    "\n",
    "    with pysam.AlignmentFile(bam_fname, \"rb\") as bam:\n",
    "        allele_counter = Counter()\n",
    "        for pileup_column in tqdm(bam.pileup(), total=16569, desc = 'consensus dna'):\n",
    "            assert pileup_column.reference_name == 'chrM'\n",
    "            pos = pileup_column.reference_pos\n",
    "\n",
    "            allele_counter.clear()\n",
    "            for pileup_read in pileup_column.pileups:\n",
    "                if pileup_read.is_del:\n",
    "                    allele = \"-\"\n",
    "                else:\n",
    "                    allele = pileup_read.alignment.query_sequence[\n",
    "                        pileup_read.query_position]\n",
    "                allele_counter[allele] += 1\n",
    "\n",
    "            max_allele = \"N\"\n",
    "            max_count, total_count = 0, 0\n",
    "            for allele, count in allele_counter.items():\n",
    "                if count > max_count:\n",
    "                    max_count = count\n",
    "                    max_allele = allele\n",
    "                total_count += count\n",
    "\n",
    "            assert max_allele in \"ACGTN-\"\n",
    "            if (max_count >= ac_threshold and\n",
    "                max_count / total_count >= af_threshold):\n",
    "                consensus += max_allele\n",
    "\n",
    "    return consensus\n",
    "\n",
    "\n",
    "@cython.cdivision(True)\n",
    "@cython.wraparound(False)\n",
    "@cython.boundscheck(False)\n",
    "def get_MN(char[:, :] genomes,str bam_fname):\n",
    "    bam = pysam.AlignmentFile(bam_fname, \"rb\")\n",
    "    cdef double[:, :] N, M\n",
    "    cdef int k, i, j, pos\n",
    "    cdef double correct, incorrect, P_cor\n",
    "    cdef str seq\n",
    "    cdef long num_reads = get_num_reads(bam_fname)\n",
    "    cdef int num_genomes = genomes.shape[0]\n",
    "    N = np.zeros((num_reads, num_genomes))\n",
    "    M = np.zeros((num_reads, num_genomes))\n",
    "    i = 0\n",
    "    j = 0\n",
    "    for read in tqdm(bam.fetch('chrM'), total = bam.count(), desc = 'MN tables'):\n",
    "        seq = read.query_sequence\n",
    "        pos = read.reference_start\n",
    "        qual = read.query_qualities\n",
    "        # print(qual)\n",
    "        if not read.is_mapped:\n",
    "            continue\n",
    "        for j in range(num_genomes):\n",
    "            correct = 0\n",
    "            incorrect = 0\n",
    "            for k in range(len(seq)):\n",
    "                if k+pos >= genomes.shape[1]:\n",
    "                    break\n",
    "                # if chr(genomes[j][k+pos]).upper() == '-':\n",
    "                #     # print('FFFFF')\n",
    "                #     M[i, j] = -1\n",
    "                #     N[i, j] = -1\n",
    "                if seq[k] == chr(genomes[j][k+pos]).upper():\n",
    "                    P_cor = 10**(- qual[k]/10)\n",
    "                    correct += P_cor\n",
    "                    incorrect += 1 - P_cor\n",
    "                else:\n",
    "                    # print(seq[k], chr(genomes[j][k+pos]).upper())\n",
    "                    P_cor = (1 - 10**(- qual[k]/10))/3\n",
    "                    incorrect += 1 - P_cor\n",
    "                    correct += P_cor\n",
    "            if M[i, j] != -1:\n",
    "                M[i, j] = correct\n",
    "                N[i, j] = incorrect\n",
    "        i += 1\n",
    "    bam.close()\n",
    "    return np.array(M, dtype=np.float64), np.array(N, dtype=np.float64)\n",
    "\n",
    "@cython.cdivision(True)\n",
    "@cython.wraparound(False)\n",
    "@cython.boundscheck(False)\n",
    "def get_mc(long[:, ::1] m, long[:, ::1] n, double eps):\n",
    "    cdef double[:,::1] mc\n",
    "    cdef long num_reads = m.shape[0]\n",
    "    cdef long num_genomes = m.shape[1] \n",
    "    cdef long j\n",
    "    mc = np.zeros((num_reads, num_genomes))\n",
    "    for i in range(num_reads):\n",
    "        for j in range(num_genomes):\n",
    "            if m[i, j] == -1:\n",
    "                mc[i, j] = 0\n",
    "            else:\n",
    "                mc[i, j] = binom(m[i, j] + n[i, j], m[i, j]) * pow((1 - eps),(m[i, j])) * pow(eps,(n[i, j]))\n",
    "    return np.asarray(mc)\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "@cython.cdivision(True)\n",
    "@cython.wraparound(False)\n",
    "@cython.boundscheck(False)\n",
    "def get_Zi(double[:,::1] mc, double[::1] p,double eps, long i):\n",
    "    cdef long num_reads = mc.shape[0]\n",
    "    cdef long num_genomes = mc.shape[1] \n",
    "    cdef long j\n",
    "    cdef long Z\n",
    "    cdef double[:] probs\n",
    "    cdef double s\n",
    "    s = 0\n",
    "    probs = np.zeros(num_genomes, dtype = float)\n",
    "    \n",
    "\n",
    "    for j in range(num_genomes):\n",
    "        probs[j] = mc[i, j] * p[j]\n",
    "        s += probs[j]\n",
    "    if s == 0:\n",
    "        return np.random.randint(0, num_genomes)\n",
    "    else:\n",
    "        for j in range(num_genomes):\n",
    "            probs[j] =probs[j] / s\n",
    "    \n",
    "    Z = np.random.choice(np.arange(0, num_genomes), p = probs)\n",
    "    # print(np.asarray(probs))\n",
    "    return Z\n",
    "\n",
    "\n",
    "\n",
    "@cython.cdivision(True)\n",
    "@cython.wraparound(False)\n",
    "@cython.boundscheck(False)\n",
    "def get_Zi_new(double[:,::1] mc, double[::1] p,double eps, long i):\n",
    "    cdef long num_reads = mc.shape[0]\n",
    "    cdef long num_genomes = mc.shape[1] \n",
    "    cdef long j\n",
    "    cdef long Z\n",
    "    cdef double[:] probs\n",
    "    cdef double s\n",
    "    cdef double rval = uniform()\n",
    "    # print(rval)\n",
    "    s = 0\n",
    "    probs = np.zeros(num_genomes, dtype = float)\n",
    "    \n",
    "\n",
    "    for j in range(num_genomes):\n",
    "        probs[j] = mc[i, j] * p[j]\n",
    "        s += probs[j]\n",
    "\n",
    "    for j in range(num_genomes):\n",
    "        probs[j] =probs[j] / s\n",
    "        \n",
    "    \n",
    "    \n",
    "    \n",
    "    cdef double cum_probs = 0\n",
    "    cdef int choice = 0\n",
    "    while True:\n",
    "        if cum_probs > rval:\n",
    "            break\n",
    "        cum_probs += probs[choice]\n",
    "        choice += 1\n",
    "    # print(np.asarray(probs))\n",
    "    return choice\n",
    "\n",
    "\n",
    "@cython.cdivision(True)\n",
    "@cython.wraparound(False)\n",
    "@cython.boundscheck(False)\n",
    "def get_eta(long[:] z,int num_genomes):\n",
    "    cdef long[:] eta\n",
    "    eta = np.zeros(num_genomes, dtype = int)\n",
    "    cdef int num_reads\n",
    "    num_reads = z.shape[0]\n",
    "    cdef int i\n",
    "    for i in range(num_reads):\n",
    "        eta[z[i]] += 1\n",
    "    return np.array(eta)\n",
    "\n",
    "cdef extern from \"stdlib.h\":\n",
    "    double drand48()\n",
    "    void srand48(long int seedval)\n",
    "\n",
    "def uniform():\n",
    "    return drand48() #This gives a float in range [0,1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fafffee6-7c80-476a-bca2-f26ca84d894e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocess(ref_fname, genomes_fname, bam_fname):\n",
    "    base = bam_fname[:-4]\n",
    "    pysam.index(bam_fname);\n",
    "    os.system(f\"samtools view {bam_fname} chrM -o {base+'_mt.bam'}\")\n",
    "    base = base+'_mt'\n",
    "    print('#EXTRACTING MTDNA OK')\n",
    "    consensus = bam2consensus(ref_fname, bam_fname)\n",
    "    consensus_fa = '>chrM\\n'+''.join(consensus) +'\\n'\n",
    "    with open(f'{base}.fa', 'w') as new_genomes:\n",
    "        new_genomes.write(consensus_fa)\n",
    "        \n",
    "    os.system(f'cat {base}.fa {genomes_fname} > {base}_genomes.fa');\n",
    "    os.system(f'mafft {base}_genomes.fa >  {base}_aligned.fa')\n",
    "    aligned_genomes = f'{base}_aligned.fa'\n",
    "    print(\"#ALL GENOMES ARE READY\")\n",
    "    new_cons = list(SeqIO.parse(f'{base}_aligned.fa', \"fasta\"))[0]\n",
    "    SeqIO.write(new_cons, f'{base}.real.fa', \"fasta\")\n",
    "    os.system(f'bwa index -a bwtsw {base}.real.fa')\n",
    "    os.system(f'samtools faidx {base}.real.fa')\n",
    "    os.system(f'rm {base}.dict')\n",
    "    os.system(f'picard CreateSequenceDictionary R={base}.fa O={base}.dict')\n",
    "    os.system(f'samtools fastq {bam_fname} > {base}.fq')\n",
    "    os.system(f'bwa aln -l 1000 -t 10 {base}.real.fa {base}.fq > {base}_ra.sai')\n",
    "    os.system(f\"bwa samse -r '@RG\\\\tID:{base}\\\\tLB:{base}_L1\\\\tPL:ILLUMINA\\\\tSM:{base}' {base}.real.fa {base}_ra.sai {base}.fq | samtools view -bh -q 30 | samtools sort -O BAM -o {base}_ra.sort.bam\")\n",
    "    os.system(f'picard MarkDuplicates I={base}_ra.sort.bam O={base}_ra.sort.rmdup.bam METRICS_FILE=metrics.txt REMOVE_DUPLICATES=true ASSUME_SORTED=false VALIDATION_STRINGENCY=LENIENT')\n",
    "    pysam.index(f'{base}_ra.sort.rmdup.bam');\n",
    "    os.system(f'samtools calmd -Erb {base}_ra.sort.rmdup.bam {base}.fa > {base}_ra.final.bam 2>/dev/null');\n",
    "    bam_final = f'{base}_ra.final.bam'\n",
    "    os.system(f'samtools index {base}_ra.final.bam')\n",
    "    os.system(f'rm {base}_ra.sai')\n",
    "    print(\"#BAM FILE IS READY\")\n",
    "    return bam_final, aligned_genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64b3bd29-c667-428f-8c2e-d8f0c348c150",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_genomes_arr(genomes_fname):\n",
    "    genomes = list()\n",
    "    for record in SeqIO.parse(genomes_fname, \"fasta\"):\n",
    "            genomes.append(str(record.seq))\n",
    "    genomes_arr = np.array([list(x) for x in genomes], dtype = 'S1')\n",
    "    return genomes_arr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c4dff22-5941-4efe-9b5d-525397f7016b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_same(genomes_arr):\n",
    "    same_positions = []\n",
    "    same_bases = []\n",
    "    for i in range(genomes_arr.shape[1]):\n",
    "        if len(np.unique(genomes_arr[:,i])) == 1:\n",
    "            same_positions.append(i)\n",
    "            same_bases.append(genomes_arr[0,i])\n",
    "    same = dict(zip( same_positions, same_bases ))\n",
    "    return same"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8076281-95a3-4f18-846c-3c122afcd7f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_base_err(bam_fname, same_dict):\n",
    "    samfile = pysam.AlignmentFile(bam_fname, \"rb\" )\n",
    "    same_positions = list(same_dict.keys())\n",
    "    correct = 0\n",
    "    total = 0\n",
    "    samfile = pysam.AlignmentFile(bam_fname, \"rb\" )\n",
    "    for pileupcolumn in tqdm(samfile.pileup(\"chrM\")):\n",
    "        pos = pileupcolumn.pos\n",
    "        if pos not in same_positions:\n",
    "            continue\n",
    "\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            if not pileupread.is_del and not pileupread.is_refskip:\n",
    "                total += 1\n",
    "                # query position is None if is_del or is_refskip is set.\n",
    "                nbase =  pileupread.alignment.query_sequence[pileupread.query_position]\n",
    "                if nbase == same[pos].decode('ascii').upper():\n",
    "                    correct += 1\n",
    "    base_err = 1 - correct/total\n",
    "    samfile.close()\n",
    "    return base_err"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa4749f2-1a7c-46a5-9741-127701459456",
   "metadata": {},
   "outputs": [],
   "source": [
    "# samfile.close()\n",
    "# del genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6b24df5-dc76-4f75-adf8-812813ba06d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def Pr_Dep(p, m, n, i, eps):\n",
    "#     num_reads, num_genomes = M.shape\n",
    "#     ret = 0\n",
    "#     for j in range(num_genomes):\n",
    "#         ret += p[j]* Pr_De(m, n, i, j, eps)\n",
    "#     return ret"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd051703-99f8-4360-8f16-4cf67cf38fe3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# def get_Z(m, n, p, eps):\n",
    "#     num_reads, num_genomes = M.shape\n",
    "#     Z = np.zeros(num_reads)\n",
    "#     for i in tqdm(range(num_reads), desc = f'gettitng Z'):\n",
    "#         probs = np.array([Pr_De(m, n, i, j, eps) * p[j] for j in range(num_genomes)])\n",
    "#         probs = probs/probs.sum()\n",
    "#         Z[i] = np.random.choice(np.arange(0, num_genomes), p = probs)\n",
    "#         # print(f'p = {p}\\nprobs={probs}\\nZ={Z[j]}')\n",
    "#     return Z"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1072486a-cd74-4339-a19b-981e9b5dd444",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def get_eta(z, num_genomes):\n",
    "#     return np.array([len(z[z==j]) for j in range(num_genomes)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b35a145-06e6-4b4e-81db-8c68fc38ceb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# p = np.array(np.random.dirichlet([1]*num_genomes)\n",
    "def do_mcmc(n_iterations = 50000, output_file = ''):\n",
    "    if output_file != '':\n",
    "        res = open(output_file,'w')\n",
    "    num_reads, num_genomes  = MC.shape\n",
    "    print(MC.shape)\n",
    "    p = np.array([0.5] + [1/(num_genomes-1)]*(num_genomes-1))\n",
    "    pool = Pool()\n",
    "    for i in tqdm(range(n_iterations) ):\n",
    "        \n",
    "        func = lambda x: get_Zi(MC, p, base_err, x)\n",
    "        \n",
    "        Z = np.array(pool.map_async(func, range(num_reads)).get())\n",
    "        \n",
    "        \n",
    "        eta = get_eta(Z, num_genomes)\n",
    "        p0 = np.random.beta(1 + eta[0],1+num_reads-eta[0])\n",
    "        p_other = np.random.dirichlet(1+ eta[1:])\n",
    "        p_other *= (1-p0)/p_other.sum()\n",
    "        p[0] = p0\n",
    "        p[1:] = p_other\n",
    "        if output_file != '':\n",
    "            res.write(f'iteration {i}')\n",
    "            res.write(str(p[0]))\n",
    "        if i % 100 == 0:\n",
    "            print(p[0], p[1:].sum()) \n",
    "    pool.close()\n",
    "    if output_file != '':\n",
    "        res.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b54e67c-cfb5-49f7-a47b-ddbb2d527798",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def do_mcmc_new(n_iterations = 50000):\n",
    "#     res = open(f'{base}_mcmc.txt','w')\n",
    "#     p = np.array([0.5] + [1/(num_genomes-1)]*(num_genomes-1))\n",
    "#     pool = Pool()    \n",
    "#     for i in tqdm(range(n_iterations), miniters=10):\n",
    "        \n",
    "#         func = lambda s : get_Zi_new(MC, p, base_err, s)\n",
    "#         # print(rvals[i])\n",
    "#         Z = np.array(pool.map_async(func, range(num_reads)).get())\n",
    "        \n",
    "        \n",
    "#         eta = get_eta(Z, num_genomes)\n",
    "#         p0 = np.random.beta(1 + eta[0],1+num_reads-eta[0])\n",
    "#         p_other = np.random.dirichlet(1+ eta[1:])\n",
    "#         p_other *= (1 - p0)/p_other.sum()\n",
    "#         p[0] = p0\n",
    "#         p[1:] = p_other\n",
    "#         # res.write(f'iteration {i}')\n",
    "#         # res.write(str(p[0]))\n",
    "#         if i % 100 == 0:\n",
    "#             print(p[0], p[1:].sum()) \n",
    "#     pool.close()\n",
    "#     res.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb98616e-8b7f-469f-ab2c-c6fb5d94007a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# p = np.array([0.5] + [1/(num_genomes-1)]*(num_genomes-1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a82e67d1-81e5-4541-b06f-76e76734529f",
   "metadata": {},
   "outputs": [],
   "source": [
    "bam, genomes = preprocess(ref_fname, genomes_fname, bam_fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef5022fa-8d41-4b0b-8f70-47698cd03d6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "genomes_arr = make_genomes_arr(genomes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48fcb8d5-78c1-484e-a736-0341bf86a173",
   "metadata": {},
   "outputs": [],
   "source": [
    "same = get_same(genomes_arr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a54fabd3-8eac-4ef8-a801-d7ac6539268a",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_err = get_base_err(bam, same)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d930dd9-6b81-4d10-b142-d3fad34c840e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e453add3-cd0b-4766-8616-e3806302a8c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "M, N = get_MN(genomes_arr, bam)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c258fd41-d58f-4fc7-984e-6e79a1b6b5cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "MC = get_mc(M, N, base_err)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5756f442-e09a-4338-954a-eb23da3fc4c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "811bb1d6-8d86-45a0-9972-120333c3a8d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "do_mcmc()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa1eb6b0-3ff9-4bcd-842b-812f09eaf3fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "p = np.array([0.5] + [1/(num_genomes-1)]*(num_genomes-1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3aee1057-c1a7-4186-b932-3e57fd148bb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "%timeit get_Zi(MC, p, num_genomes, 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afca0e62-9906-4f6f-8051-78202763d463",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a933456-a3f5-440c-aef2-82eb7b7033c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2210925-485b-4e5a-91b4-b75a6259f43d",
   "metadata": {},
   "outputs": [],
   "source": [
    "bam_1 = pysam.AlignmentFile(bam_fname, \"rb\")\n",
    "list(bam.fetch('chrM'))[0].seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb4d0621-a2f5-4740-a473-06dee9349b63",
   "metadata": {},
   "outputs": [],
   "source": [
    "10**(-list(bam_1.fetch('chrM'))[0].query_qualities[-1]/10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6384c4f0-32ce-4031-8096-2de80cba89eb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
